### PR TITLE
Goals: Schedule multi month forecasting

### PR DESCRIPTION
**Pull request by shall0pass**
References discord discussion starting here: https://discord.com/channels/937901803608096828/940290142579605514/1133523705063030824

Currently the schedule keyword won't fill any future budget cells if the category balance already satisfies the schedule.  This PR is an attempt to improve the behavior by allowing budget fills regardless of the category balance.

This is a drastic rewrite of the schedule keyword.  Though I've tried not to have any regressions, it is possible because of how different the logic is.  I've tested compounding using a simple template, so a small change in the 'by' keyword was also made.

### CODE REVIEW

**Comment by kyrias** [comment]
I've done some testing of this now and it seems to work exactly like I expected it to work!  :)

**Comment by youngcw** [comment]
Using the full flag seems broken.  It will work well when the previous months have fully allocated funds, but if the category is at 0 then it will act like its a regular schedule.

**Comment by shall0pass** [comment]
> Using the full flag seems broken. It will work well when the previous months have fully allocated funds, but if the category is at 0 then it will act like its a regular schedule.

Thanks for checking.  I'll try to fix that up.

**Comment by youngcw** [comment]
something broke with my priority test.  It ends up getting negative values if the available funds are negative at time of priority pass (I think)

**Comment by shall0pass** [comment]
> something broke with my priority test. It ends up getting negative values if the available funds are negative at time of priority pass (I think)

I think I have it resolved.  Could you run your test again?  Thanks!

**Comment by youngcw** [comment]
Its a little better.  It works well if started in the right month, but still budgets the whole thing in wrong months when the balance is 0 when applying the template.

**Comment by joel-jeremy** [comment]
Haven't done an in depth review of the code yet but the current implementation for schedule templates allocate a budget even though the schedule is already completed. Is it still the same here? I think we change that and just ignore the template if it's schedule is already completed. Thoughts?

**Comment by shall0pass** [comment]
Currently it's the same.  I'll look at doing a check for a completed schedule though, that's a great point.

I'm still checking for bugs. I found a few more tonight that I hope I resolved with the latest commit, but I don't think I've addressed @youngcw comment about the negative value in a priority level.  I'm still trying to replicate the conditions for that to occur so I can fix it.

**Review by youngcw — CHANGES_REQUESTED** [review]

File: packages/loot-core/src/server/budget/goaltemplates.ts [hunk 1]
```diff
@@ -411,10 +410,10 @@ async function applyCategoryTemplate(
         } else {
           increment = limit;
         }
-        if (increment < budgetAvailable || !priority) {
+        if (to_budget + increment < budgetAvailable || !priority) {
           to_budget += increment;
         } else {
-          if (budgetAvailable > 0) to_budget += budgetAvailable;
```
[the remark this answers is not in the document]
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goaltemplates.ts:417]: This was part of it. The other piece that was off is above.  It was passing two different priority level templates at once when schedules were involved.  So a 'by' template with a priority of 0 and a schedule with a priority of 2 would get run at the same time.  Obviously that wasn't intended.

**Comment by shall0pass** [comment]
I've also noticed that I'm not getting error or warning pop ups.  The errors still seem to be passed back to the calling function correctly.  Was there a change in the notification system that we need to address from this file?

**Comment by youngcw** [comment]
Priority thing is fixed.  With the full flag it seems to be that after 3 broken months future months fill right.  I can shift which month I start and it looks to be the same, 3 broken then good.  Maybe that's just an artifact of my test.

**Comment by shall0pass** [comment]
I think I broke the error notifications when doing the speed up improvements.  The functions are async and those are the only functions not passing the Notifications back.  The Check templates and End of month cleanup notifications still come up.  I've been trying things, but I'm poking at something I don't fully understand.  Could use a hint. 🙏

**Comment by youngcw** [comment]
Beautiful.  All my tests are working now.

I can't tell what happened to the regular messages, but the messages for the template check still work.

**Comment by MatissJanis** [comment]
@youngcw would you mind doing a code review here? I could do it myself, but since I don't really use the feature.. it would be mostly a rubber stamp.

**Review by youngcw — APPROVED** [review]
Thanks for figuring this out.

**Comment by joel-jeremy** [comment]
Looks good! Tested and it no longer applies completed schedules. Thanks!

**Review by joel-jeremy — APPROVED** [review]

**Pull request merged**